src/Microsoft.Management.Configuration/Database/Schema/IConfigurationDatabase.cpp

```diff
@@ -4,34 +4,74 @@
+        {
+            using StorageT = std::shared_ptr<AppInstaller::SQLite::SQLiteDynamicStorage>;
+
+            if (version.MajorVersion == 0)
+            {
+                constexpr std::array<std::unique_ptr<IConfigurationDatabase>(*)(const StorageT& s), 2> versionCreatorMap =
+                {
+                    [](const StorageT& s) { return std::unique_ptr<IConfigurationDatabase>(std::make_unique<Database::Schema::V0_1::Interface>(s)); },
+                    [](const StorageT& s) { return std::unique_ptr<IConfigurationDatabase>(std::make_unique<Database::Schema::V0_2::Interface>(s)); },
+                };
+
+                size_t minorVersion = static_cast<size_t>(version.MinorVersion);
+                if (minorVersion >= 1 && minorVersion <= versionCreatorMap.size())
+                {
+                    return versionCreatorMap[minorVersion - 1](storage);
+                }
+            }
+
+            // We do not have the capacity to operate on this schema version
+            THROW_WIN32(ERROR_NOT_SUPPORTED);
+        }
+    }
+
     AppInstaller::SQLite::Version IConfigurationDatabase::GetLatestVersion()
     {
-        return { 0, 1 };
+        return { 0, 2 };
     }
 
-    std::unique_ptr<IConfigurationDatabase> IConfigurationDatabase::CreateFor(std::shared_ptr<AppInstaller::SQLite::SQLiteDynamicStorage> storage)
+    std::unique_ptr<IConfigurationDatabase> IConfigurationDatabase::CreateFor(const std::shared_ptr<AppInstaller::SQLite::SQLiteDynamicStorage>& storage, bool allowMigration)
     {
         using StorageT = std::shared_ptr<AppInstaller::SQLite::SQLiteDynamicStorage>;
         const AppInstaller::SQLite::Version& version = storage->GetVersion();
 
-        if (version.MajorVersion == 0)
-        {
-            constexpr std::array<std::unique_ptr<IConfigurationDatabase>(*)(StorageT&& s), 1> versionCreatorMap =
-            {
-                [](StorageT&& s) { return std::unique_ptr<IConfigurationDatabase>(std::make_unique<Database::Schema::V0_1::Interface>(std::move(s))); },
-            };
+        std::unique_ptr<IConfigurationDatabase> result = CreateForVersion(version, storage);
 
-            size_t minorVersion = static_cast<size_t>(version.MinorVersion);
-            if (minorVersion >= 1 && minorVersion <= versionCreatorMap.size())
-            {
-                return versionCreatorMap[minorVersion - 1](std::move(storage));
-            }
+        AppInstaller::SQLite::Version latestVersion = GetLatestVersion();
+        if (allowMigration && version < latestVersion)
+        {
+            // Always migrate to the latest version until a reason comes along to not do that
+            std::unique_ptr<IConfigurationDatabase> latest = CreateForVersion(latestVersion, storage);
+            THROW_WIN32_IF(ERROR_NOT_SUPPORTED, !latest->MigrateFrom(result.get()));

```

Comment:
Did we consider falling back to use previous version if migration failed? I saw the sequencer supports "database does not support queueing"

Reply:
I think that is probably reasonable past a 1.0 schema, especially if there is some actual potential for failure / data loss that we want to avoid.  The comment is intended to indicate that the hard failure isn't required forever.